From 7b14042c90ff8d11ddfedd7ac8393464366cedeb Mon Sep 17 00:00:00 2001
From: Julien Gilli <julien.gilli@joyent.com>
Date: Wed, 4 May 2016 12:22:19 +0200
Subject: [PATCH] TOOLS-1857 Add sdc-volapi command to headnode's GZ TOOLS-1505
 Make sdc-volapi error with a clear error message if experimental support for
 NFS shared volumes is not enabled TOOLS-1824 sdcadm update-other should not
 do anything for VOLAPI Reviewed by: Trent Mick <trentm@gmail.com> Approved
 by: Trent Mick <trentm@gmail.com>

---
 CHANGES.md                  |  4 ++++
 bin/sdc-volapi              | 44 +++++++++++++++++++++++++++++++++++++
 lib/libdc.sh                | 20 ++++++++++++++++-
 package.json                |  2 +-
 sapi_manifests/sdc/template |  3 ++-
 5 files changed, 70 insertions(+), 3 deletions(-)
 create mode 100755 bin/sdc-volapi

diff --git a/CHANGES.md b/CHANGES.md
index ecf34fb..12409e9 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,5 +1,9 @@
 # sdc (SDC ops core zone) Changelog
 
+## 1.5.0
+
+- TOOLS-1857 Add sdc-volapi command to headnode's GZ
+
 ## 1.4.0
 
 - Manta instances with a large number of load balancer IP addresses are
diff --git a/bin/sdc-volapi b/bin/sdc-volapi
new file mode 100755
index 0000000..a16b22a
--- /dev/null
+++ b/bin/sdc-volapi
@@ -0,0 +1,44 @@
+#!/usr/bin/bash
+#
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+
+#
+# Copyright (c) 2014, Joyent, Inc.
+#
+
+#
+# Convenience wrapper for calling VOLAPI.
+#
+
+if [ "$TRACE" != "" ]; then
+    export PS4='[\D{%FT%TZ}] ${BASH_SOURCE}:${LINENO}: ${FUNCNAME[0]:+${FUNCNAME[0]}(): }'
+    set -o xtrace
+fi
+set -o errexit
+set -o pipefail
+
+TOP=$(cd $(dirname $0)/../ 2>/dev/null; pwd)
+LIBDC=$TOP/lib/libdc.sh
+if [[ ! -f $LIBDC ]]; then
+    echo "$(basename $0): fatal: unable to find $LIBDC"
+    exit 1
+fi
+source $LIBDC
+
+path=$1
+if [[ -z ${path} ]]; then
+    cat <<EOF
+error: no PATH argument given
+
+Usage:
+        $0 [--no-headers] PATH [EXTRA-CURL-ARGS]
+
+EOF
+    exit 1
+fi
+
+volapi "$@" | (json -q || true)
+exit $?
diff --git a/lib/libdc.sh b/lib/libdc.sh
index e301d37..2c24073 100755
--- a/lib/libdc.sh
+++ b/lib/libdc.sh
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2015, Joyent, Inc.
+# Copyright (c) 2017, Joyent, Inc.
 #
 
 #
@@ -205,6 +205,24 @@ function mahi() {
     return 0
 }
 
+VOLAPI_URL=
+function volapi() {
+    local path=$1
+    shift
+    if [[ -z ${VOLAPI_URL} ]]; then
+        VOLAPI_DOMAIN=$(json -f ${CONFIG} volapi_domain)
+        [[ -n ${VOLAPI_DOMAIN} ]] \
+            || fatal "fatal: volapi_domain not defined in config"
+        VOLAPI_URL="http://${VOLAPI_DOMAIN}"
+    fi
+    if [[ -z "$SDC_API_VERSION" ]]; then
+        SDC_API_VERSION="*"
+    fi
+    (curl ${CURL_OPTS} -H "accept-version: ${SDC_API_VERSION}" \
+        --url "${VOLAPI_URL}${path}" "$@") || return $?
+    echo ""  # sometimes the result is not terminated with a newline
+    return 0
+}
 
 # filename passed must have a 'Job-Location: ' header in it.
 watch_job()
diff --git a/package.json b/package.json
index aeb43af..dbf2f8b 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
     "name": "sdc",
     "description": "SmartDataCenter 'sdc' zone to hold ops/admin tools",
-    "version": "1.4.0",
+    "version": "1.5.0",
     "author": "Joyent (joyent.com)",
     "private": true,
     "dependencies": {
diff --git a/sapi_manifests/sdc/template b/sapi_manifests/sdc/template
index a171415..9f4684f 100644
--- a/sapi_manifests/sdc/template
+++ b/sapi_manifests/sdc/template
@@ -28,5 +28,6 @@
     "ufds_remote_ldap_root_pw": "{{{ufds_remote_ldap_root_pw}}}",
     {{/ufds_remote_ldap_root_pw}}
     "vmapi_domain": "{{{vmapi_domain}}}",
-    "workflow_domain": "{{{workflow_domain}}}"
+    "workflow_domain": "{{{workflow_domain}}}",
+    "volapi_domain": "volapi.{{{datacenter_name}}}.{{{dns_domain}}}"
 }
-- 
2.21.0

